.github/general_coding_instructions.md

```diff
@@ -0,0 +1,798 @@
+
+### Function Organization
+- **Parameter validation first**: Always validate parameters at the beginning of the function
+- **Variable declarations**: Declare variables when first needed (C99+ style) to improve readability and reduce scope
+- **Single exit point**: Always use goto for cleanup and maintain exactly one return statement per function
+- **Error handling**: Use consistent result variable patterns for error propagation
+
+### Function Length and Complexity
+- **Keep functions focused**: Each function should have a single, well-defined responsibility
+- **Limit function length**: Prefer functions under 100 lines; split complex functions into helpers
+- **Minimize nesting**: Use goto patterns to reduce nesting levels and maintain single exit point
+- **Use helper functions**: Extract common logic into appropriately named internal helper functions
+
+### Documentation and Comments
+- **Function purpose**: Document what the function does, not how it does it
+- **Complex logic**: Add inline comments for non-obvious algorithms or business logic
+
+### Async Function Callback Patterns {#async-callbacks}
+Asynchronous operations must follow consistent callback patterns for error handling and result delivery:
+
+```c
+// Standard async callback signature
+typedef void(*MY_ASYNC_CALLBACK)(
+    void* context,                    // User-provided context
+    MY_RESULT result,                 // Operation result (success/failure)
+    /* ... additional result data ... */
+);
+
+// Async function signature pattern
+int my_operation_async(
+    MY_HANDLE handle,
+    const char* param,
+    MY_ASYNC_CALLBACK callback,       // Callback function
+    void* callback_context            // User context passed to callback
+);
+```
+
+**Async Callback Rules:**
+- **Error vs Success behavior**: If async function returns error, callback is NEVER called. If async function returns success, callback is called exactly ONCE
+- **Single callback invocation**: Each successfully started async operation calls the callback exactly once
+- **Context preservation**: Pass user context through unchanged to callback
+- **Error reporting**: Use result parameter to indicate success/failure
+- **Thread safety**: Callbacks may be invoked from any thread
+
+**Async Implementation Pattern:**
+```c
+int my_operation_async(MY_HANDLE handle, const char* param, MY_ASYNC_CALLBACK callback, void* context)
+{
+    int result;
+    
+    // Parameter validation
+    if (handle == NULL || callback == NULL)
+    {
+        LogError("Invalid arguments");
+        result = MU_FAILURE;
+    }
+    else
+    {
+        // Start async work
+        if (start_async_work(handle, param, callback, context) == 0)
+        {
+            result = 0;  // Async operation started successfully
+            goto callback_will_come;
+        }
+        else
+        {
+            result = MU_FAILURE;  // Failed to start async operation
+        }
+    }
+    
+callback_will_come:
```

Comment:
>callback_will_come [](http://example.com/codeflow?start=0&length=18)

This is may be descriptive, but all_ok is what we use.

Reply:
Usually when a callback is expected Andrei uses callback_will_come. And I've seen others use that too.